### PR TITLE
Switch browser tests to playwright

### PR DESCRIPTION
**Pull request by alxndrsn**
# Motivation

* I can't get Selenium tests running locally
* I managed to port the tests to Playwright very easily
* I've found playwright very reliable

# Included

* remove saucelabs support from browser tests: this doesn't seem to be used elsewhere (PR at https://github.com/pouchdb/pouchdb/pull/8577)
* re-introduce test referenced in https://github.com/pouchdb/pouchdb/issues/6835

### CODE REVIEW

**Comment by alxndrsn** [comment]
Current failures:
```

  browser.worker.js
    1) "before each" hook
    2) "after all" hook


  1708 passing (3m)
  41 pending
  2 failing

  1) browser.worker.js "before each" hook:
     Uncaught uncaught exception: AssertionError: expected 0 to equal 7 (http://127.0.0.1:8000/node_modules/chai/chai.js:206)
  startTests@webrunner.js:82:11
  EventListener.handleEvent*@webrunner.js:90:14
  promise callback*@webrunner.js:85:27
  @webrunner.js:94:3
  

  2) browser.worker.js "after all" hook:
     Uncaught uncaught exception: AssertionError: expected 0 to equal 7 (http://127.0.0.1:8000/node_modules/chai/chai.js:206)
  startTests@webrunner.js:82:11
  EventListener.handleEvent*@webrunner.js:90:14
  promise callback*@webrunner.js:85:27
  @webrunner.js:94:3
  ```
  and
  ```
  Circular dependency: packages/node_modules/pouchdb-collections/lib/index.es.js -> packages/node_modules/pouchdb-collections/lib/index.es.js
  ✓ wrote pouchdb-browser/lib/index.js in vanilla mode
  ✓ wrote pouchdb-browser/lib/index.es.js in vanilla mode
  ✓ wrote pouchdb-node/lib/index.js in vanilla mode
  ✓ wrote pouchdb-node/lib/index.es.js in vanilla mode
build error
Error: 'default' is not exported by node_modules/inherits/inherits.js
    at error (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:3460:30)
    at Module.error (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:13349:9)
    at handleMissingExport (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:13039:21)
    at Module.traceVariable (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:13453:17)
    at ModuleScope.findVariable (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:12726:29)
    at Identifier$$1.bind (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:10111:40)
    at CallExpression.NodeBase.bind (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:9183:23)
    at CallExpression.bind (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:10224:31)
    at ExpressionStatement$$1.NodeBase.bind (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:9183:23)
    at Program.NodeBase.bind (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:9179:31)
    at Module.bindReferences (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:13303:18)
    at Graph.link (/home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:21489:20)
    at /home/runner/work/pouchdb/pouchdb/node_modules/rollup/dist/rollup.js:21584:19
    at async Promise.all (index 0)
    at async Promise.all (index 20)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pouchdb-monorepo@7.0.0-prerelease build-modules: `node bin/build-modules.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the pouchdb-monorepo@7.0.0-prerelease build-modules script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2023-04-26T09_56_02_740Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pouchdb-monorepo@7.0.0-prerelease build: `npm run build-modules && npm run build-test`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the pouchdb-monorepo@7.0.0-prerelease build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2023-04-26T09_56_02_756Z-debug.log
npm WARN acorn-import-assertions@1.8.0 requires a peer of acorn@^8 but none is installed. You must install peer dependencies yourself.

+ webpack-cli@4.9.2
+ webpack@5.66.0
updated 2 packages and audited 1079 packages in 4.641s

82 packages are looking for funding
  run `npm fund` for details

found 99 vulnerabilities (8 low, 32 moderate, 47 high, 12 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
assets by status 0 bytes [cached] 1 asset

ERROR in main
Module not found: Error: Can't resolve './packages/node_modules/pouchdb' in '/home/runner/work/pouchdb/pouchdb'
resolve './packages/node_modules/pouchdb' in '/home/runner/work/pouchdb/pouchdb'
  using description file: /home/runner/work/pouchdb/pouchdb/package.json (relative path: .)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: /home/runner/work/pouchdb/pouchdb/package.json (relative path: ./packages/node_modules/pouchdb)
      no extension
        Field 'browser' doesn't contain a valid alias configuration
        /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb is not a file
      .js
        Field 'browser' doesn't contain a valid alias configuration
        /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb.js doesn't exist
      .json
        Field 'browser' doesn't contain a valid alias configuration
        /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb.json doesn't exist
      .wasm
        Field 'browser' doesn't contain a valid alias configuration
        /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb.wasm doesn't exist
      as directory
        existing directory /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb
          using description file: /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json (relative path: .)
            use ./lib/index.es.js from module in package.json
              using description file: /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json (relative path: ./lib/index.es.js)
                no extension
                  aliased from description file /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json with mapping './lib/index.es.js' to './lib/index-browser.es.js'
                    using description file: /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json (relative path: .)
                      using description file: /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json (relative path: ./lib/index-browser.es.js)
                        no extension
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.es.js doesn't exist
                        .js
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.es.js.js doesn't exist
                        .json
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.es.js.json doesn't exist
                        .wasm
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.es.js.wasm doesn't exist
                        as directory
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.es.js doesn't exist
                .js
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index.es.js.js doesn't exist
                .json
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index.es.js.json doesn't exist
                .wasm
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index.es.js.wasm doesn't exist
                as directory
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index.es.js doesn't exist
            use ./lib/index.js from main in package.json
              using description file: /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json (relative path: ./lib/index.js)
                no extension
                  aliased from description file /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json with mapping './lib/index.js' to './lib/index-browser.js'
                    using description file: /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json (relative path: .)
                      using description file: /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json (relative path: ./lib/index-browser.js)
                        no extension
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.js doesn't exist
                        .js
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.js.js doesn't exist
                        .json
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.js.json doesn't exist
                        .wasm
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.js.wasm doesn't exist
                        as directory
                          /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index-browser.js doesn't exist
                .js
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index.js.js doesn't exist
                .json
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index.js.json doesn't exist
                .wasm
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index.js.wasm doesn't exist
                as directory
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index.js doesn't exist
            using path: /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/index
              using description file: /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json (relative path: ./index)
                no extension
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/index doesn't exist
                .js
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/index.js doesn't exist
                .json
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/index.json doesn't exist
                .wasm
                  /home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/index.wasm doesn't exist

webpack 5.66.0 compiled with 1 error in 152 ms

> pouchdb-monorepo@7.0.0-prerelease test /home/runner/work/pouchdb/pouchdb
> bash bin/run-test.sh


> pouchdb-monorepo@7.0.0-prerelease build-node /home/runner/work/pouchdb/pouchdb
> bash bin/build-node.sh

internal/modules/cjs/loader.js:357
      throw err;
      ^

Error: Cannot find module '/home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/lib/index.js'. Please verify that the package.json has a valid "main" entry
    at tryPackage (internal/modules/cjs/loader.js:349:19)
    at Function.Module._findPath (internal/modules/cjs/loader.js:562:18)
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:917:27)
    at Function.Module._load (internal/modules/cjs/loader.js:774:27)
    at Module.require (internal/modules/cjs/loader.js:1003:19)
    at require (internal/modules/cjs/helpers.js:107:18)
    at Object.<anonymous> (/home/runner/work/pouchdb/pouchdb/tests/misc/pouchdb-express-router.js:8:15)
    at Module._compile (internal/modules/cjs/loader.js:1114:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1143:10)
    at Module.load (internal/modules/cjs/loader.js:979:32) {
  code: 'MODULE_NOT_FOUND',
  path: '/home/runner/work/pouchdb/pouchdb/packages/node_modules/pouchdb/package.json',
  requestPath: '../../packages/node_modules/pouchdb'
}
Waiting for host to start .....
Host failed to start
npm ERR! Test failed.  See above for more details.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pouchdb-monorepo@7.0.0-prerelease test-webpack: `bash bin/test-webpack.sh`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the pouchdb-monorepo@7.0.0-prerelease test-webpack script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2023-04-26T09_56_29_529Z-debug.log
Error: Process completed with exit code 1.
```

File: package.json
Comment (lucidNTR) [inline package.json:98]: playwright is 1.33 now and please hard pin the version as for the others..

**Comment by lucidNTR** [comment]
also could you pls rebase, i see the saucelab rmoval was moved in independently allready...

**Comment by alxndrsn** [comment]
Blocked by:

* #8651 
* #8652

**Comment by alxndrsn** [comment]
Depends on: https://github.com/pouchdb/pouchdb/pull/8682

File: bin/test-browser.js
Comment (lucidNTR) [inline bin/test-browser.js:152]: is the fixme still relevant? it seems you select the browser engine from the import not via options...

**Review by lucidNTR — CHANGES_REQUESTED** [review]
Cant wait to get this merged, the main point is probably to merge the latest version while we are at it

File: bin/test-browser.js
Comment (alxndrsn) [reply bin/test-browser.js:152]: Good catch :+1:

File: bin/test-browser.js
Comment (alxndrsn) [reply bin/test-browser.js:143]: Playwright has a class `Browser`.  `browserImpl.launch()` returns an instance of `Browser` at line 152, which is assigned to `const browser`, so that name's already taken.

Following the examples in the docs, calling the `Browser` instance `browser` seems to be idiomatic in playwright code.

+1 for `browserName` - this seems to be the standard name in e.g. https://playwright.dev/docs/api/class-fixtures#fixtures-browser-name

Any alternative suggestions to `browserImpl`?

File: bin/test-browser.js
Comment (alxndrsn) [reply bin/test-browser.js:17]: Agreed it's cryptic.  The aim is to create a whiltelist of allowed values.  In theory, you could get some weird behaviour by just allowing `const browserImpl = require('playwright')[process.env.CLIENT]`:

```js
> Object.keys(require('playwright'))
[
  '_events',
  '_eventsCount',
  '_maxListeners',
  '_connection',
  '_parent',
  '_objects',
  '_type',
  '_guid',
  '_channel',
  '_initializer',
  '_logger',
  '_instrumentation',
  '_eventToSubscriptionMapping',
  '_android',
  '_electron',
  'chromium',
  'firefox',
  'webkit',
  'devices',
  'selectors',
  'request',
  'errors',
  '_toImpl'
]
```

If the whitelist is wanted, in theory it should be possible to check:
```js
if (!(browserImpl instanceof BrowserType)) {
```
This has the benefit of being clearer and more future-proof, but I can't see a way to get a reference to the `BrowserType` class.

Would this be clearer:
```js
const playwright = require('playwright');
...
if (![ 'chromium', 'firefox', 'webkit' ].includes(browserName)) {
```
?

Comment (lucidNTR) [reply bin/test-browser.js]: ah good, then my suspicion was correct, if you ask me, i much prefer your whitelist approach as this is more explicit and understandable. i would add also an explicit error message if the whitelist check fails to say sorry browser x is not supported yet or has a command line arg typo.

File: bin/test-browser.js
Comment (lucidNTR) [reply bin/test-browser.js]: a js convention seems to be to have capitals for the class and small case for the instance. that would be my favorite, but ima also not opposed to use something like browserInstance or browserInst.

`const browser = Browser.launch()`

Comment (alxndrsn) [reply bin/test-browser.js]: I've eliminated the intermediate variable, so now there is just `browserName` and `browser`.

File: bin/test-browser.js
Comment (alxndrsn) [reply bin/test-browser.js]: Sample output:
```
$ CLIENT=chrome npm run test
...

    !!! Requested browser not supported: 'chrome'.
    !!! Available browsers: 'chromium', 'firefox', 'webkit'.

$ echo $?
1
```

**Review by lucidNTR — APPROVED** [review]
<3

From my side this looks mergeable now! unfortunately i still do not have real approval rights, should we ping someone on slack?

**Comment by garethbowen** [comment]
@lucidNTR @alxndrsn I'm happy to merge this, but because it changes the names of required builds, we need to update the branch protection rules too. Doing that will mean every other outstanding PR will need to be re-run to use these new required builds. Are you happy for me to do that now, or are there other PRs that should get merged first?

**Review by garethbowen — APPROVED** [review]

**Comment by garethbowen** [comment]
@alxndrsn I'm getting some errors after merging this to `master` but it's not failing the build. Could you have a look? https://github.com/pouchdb/pouchdb/actions/runs/5650289637

**Comment by lucidNTR** [comment]
@alxndrsn  this seems to be timeouts that we did not catch before as we could never run the test suite with the old branch protection/CI rules, am i right or reading this wrong? it might be enough to increase some timeout as we just use playwright with defaults at the moment which might be slightly lower than the custom timeouts that were be set before

**Comment by alxndrsn** [comment]
@garethbowen do you mean these errors in the github action:

Screenshot_2023-07-25_08-00-33

?

We have:

### `test.issue8581.js #8581 Should not raise error:`

* [pouchdb-server (firefox, npm test)](https://github.com/pouchdb/pouchdb/actions/runs/5650289637/job/15306436535#logs)
* [couchdb-browser (3.1, firefox, npm test)](https://github.com/pouchdb/pouchdb/actions/runs/5650289637/job/15306427098#logs)
* [browser-adapter (firefox, memory, npm test)](https://github.com/pouchdb/pouchdb/actions/runs/5650289637/job/15306428708#logs)

### `test.sync.js-http-http 5007 sync 2 databases:`

* [nodejs (16, CLIENT=firefox npm run test-webpack)](https://github.com/pouchdb/pouchdb/actions/runs/5650289637/job/15306433106#logs)
* [pouchdb-server (chromium, npm test)](https://github.com/pouchdb/pouchdb/actions/runs/5650289637/job/15306436965#logs)
* [browser-adapter (firefox, indexeddb, npm test)](https://github.com/pouchdb/pouchdb/actions/runs/5650289637/job/15306428382#logs)
* [browser-adapter (firefox, indexeddb, npm test)](https://github.com/pouchdb/pouchdb/actions/runs/5650289637/job/15306428382#logs)

These are new ones on me too :thinking:

**Comment by garethbowen** [comment]
My best guess at what's happening is the test times out on the first run through which generates the error message in the screenshot. However it then runs a retry which doesn't error, so overall the build passes.

There's a problem though - as far as I can tell the retry doesn't execute anything... From the logs:

```
// lots of errors

  [29]</</</testUtils.cleanup@utils-bundle.js:15383:7
  @test.issue3646.js:15:17
  



npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pouchdb-monorepo@7.0.0-prerelease test-browser: `npm run build-test && node ./bin/test-browser.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the pouchdb-monorepo@7.0.0-prerelease test-browser script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2023-07-24T22_34_18_786Z-debug.log
npm ERR! Test failed.  See above for more details.
Error: Process completed with exit code 1.

// First retry

Run npm test
  npm test
  shell: /usr/bin/bash -e {0}
  env:
    NODE_VERSION: 14
    TEST_HOST: localhost
    CLIENT: firefox
    SERVER: pouchdb-server

> pouchdb-monorepo@7.0.0-prerelease test /home/runner/work/pouchdb/pouchdb
> ./bin/run-test.sh

// Cleanup
```

Between the `./bin/run-test.sh` and the `// Cleanup` it should actually run the tests...

**Comment by garethbowen** [comment]
@lucidNTR @alxndrsn Ok so this has been happening in master for a while now... here's one from 7 months ago. https://github.com/pouchdb/pouchdb/actions/runs/3703850265

So it's not caused by this PR. Let's raise a new issue for it...

**Comment by garethbowen** [comment]
Issue raised: #8713

**Pull request merged**
